libraries/AP_Camera/AP_Camera_Backend.cpp

```diff
@@ -241,6 +250,133 @@ void AP_Camera_Backend::send_camera_settings(mavlink_channel_t chan) const
         NaN);               // focusLevel float, percentage from 0 to 100, NaN if unknown
 }
 
+// calculate poi information
+void AP_Camera_Backend::calculate_poi()
+{
+#if AP_TERRAIN_AVAILABLE
+    if (!poi_calculation.sem.take_nonblocking()) {
+        return;
+    }
+    const AP_AHRS &ahrs = AP::ahrs();
+    if (!ahrs.get_location(poi_calculation.loc)) {
+        gcs().send_text(MAV_SEVERITY_ERROR, "POI: vehicle pos unavailable");
+        return;
+    }
+
+    // change vehicle location to ASML
+    poi_calculation.loc.change_alt_frame(Location::AltFrame::ABSOLUTE);
+
+    // retrieve gimbal attitude
+    auto *mount = AP_Mount::get_singleton();
+    Vector3f att;
+    if (!mount->get_attitude_euler(get_mount_instance(),att[0], att[1], att[2]) || isnan(att[1]) || isnan(att[2])) {
+        gcs().send_text(MAV_SEVERITY_ERROR, "POI: gimbal attitude unavailable");
+    }
+
+    // project forward from vehicle looking for terrain
+    // start testing at vehicle's location
+    Location test_loc = poi_calculation.loc, prev_test_loc = poi_calculation.loc;
+
+    // get terrain altitude (asml) at test_loc
+    bool success = false;
+    auto terrain = AP_Terrain::get_singleton();
+    float terrain_amsl_m, prev_terrain_amsl_m;
+    success = terrain->height_amsl(test_loc, terrain_amsl_m, true);
+    prev_terrain_amsl_m = terrain_amsl_m;
+
+    if (!success) {
+        gcs().send_text(MAV_SEVERITY_ERROR, "POI: failed to get terrain alt");
+        return;
+    }
+
+    float mount_pitch_deg = att[1];
+    float mount_yaw_ef_deg = wrap_180(att[2] + degrees(ahrs.get_yaw()));
+
+    float dist_increment_m;
+    terrain->get_terrain_spacing(dist_increment_m);
+    float dist_max = 10000; // Hardcoded for testing
+    float total_dist = 0;
+
+    // iteratively move test_loc forward until its alt-above-sea-level is below terrain-alt-above-sea-level
+    while (total_dist < dist_max && (test_loc.alt * 0.01) > terrain_amsl_m) {
+        total_dist += dist_increment_m;
+
+        // Take backup of previous test location and terrain amsl
+        prev_test_loc = test_loc;
+        prev_terrain_amsl_m = terrain_amsl_m;
+
+        // Move test location forward
+        test_loc.offset_bearing_and_pitch(mount_yaw_ef_deg, mount_pitch_deg, dist_increment_m);
+
+        // Get terrain's alt-above-sea-level (at test_loc)
+        success = terrain->height_amsl(test_loc, terrain_amsl_m, true);
+
+        // Fail if terrain alt cannot be retrieved
+        if (!success || std::isnan(terrain_amsl_m)) {
```

Comment:
If this fails, continue which should lead to a delay of 0.1 sec or something before the calculation is restarted with a new vehicle position, camera attitude, etc.